staging/src/k8s.io/legacy-cloud-providers/aws/aws.go

```diff
@@ -3860,21 +3860,33 @@ func (c *Cloud) EnsureLoadBalancer(ctx context.Context, clusterName string, apiS
 		}
 	} else {
 		klog.V(4).Infof("service %v does not need custom health checks", apiService.Name)
-		// We only configure a TCP health-check on the first port
-		var tcpHealthCheckPort int32
-		for _, listener := range listeners {
-			if listener.InstancePort == nil {
-				continue
-			}
-			tcpHealthCheckPort = int32(*listener.InstancePort)
-			break
-		}
 		annotationProtocol := strings.ToLower(annotations[ServiceAnnotationLoadBalancerBEProtocol])
 		var hcProtocol string
+		var tcpHealthCheckPort int32
 		if annotationProtocol == "https" || annotationProtocol == "ssl" {
 			hcProtocol = "SSL"
+			for _, listener := range listeners {
+				if listener.InstancePort == nil {
+					continue
+				}
+				protocol := listener.Protocol
+				if *protocol == "HTTPS" || *protocol == "SSL" {
+					tcpHealthCheckPort = int32(*listener.InstancePort)
+					break
+				}
+				if tcpHealthCheckPort == 0 {
+					return nil, fmt.Errorf("Failed to assign SSL heath check port for load balancer %v, no secure endpoint available", loadBalancerName)
+				}
+			}
 		} else {
 			hcProtocol = "TCP"
+			for _, listener := range listeners {
```

Comment:
In this case, when the annotation (Backend) protocol is not HTTPS or SSL (meaning it is HTTP or TCP) , we pick the first listener that has an instance port.  Should we instead be picking the first listener that does not have HTTPS or SSL as its protocol?

Reply:
This depends - is it safe to assume that if the backend protocol is not HTTPS/SSL then there should be no HTTPS/SSL endpoints on the loadbalancer?

Reply:
No that's not a safe assumption, there can be frontend SSL listeners with TCP backend listeners if you want SSL termination at the load balancer.  But I think what you have makes sense, if we have decided the health check protocol is TCP, then it will work on any open port.